src/number_theory/cyclotomic/primitive_roots.lean

```diff
@@ -0,0 +1,199 @@
+/-
+Copyright (c) 2022 Riccardo Brasca. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alex Best, Riccardo Brasca, Eric Rodriguez
+-/
+
+import number_theory.cyclotomic.basic
+import ring_theory.adjoin.power_basis
+import ring_theory.norm
+
+/-!
+# Primitive roots in cyclotomic fields
+If `is_cyclotomic_extension {n} A B`, we define an element `zeta n A B : B` that is (under certain
+assumptions) a primitive `n`-root of unity in `B` and we study its properties. We also prove related
+theorems under the more general assumption of just being a primitive root, for reasons described
+in the implementation details section.
+
+## Main definitions
+* `is_cyclotomic_extension.zeta n A B`: if `is_cyclotomic_extension {n} A B`, than `zeta n A B`
+  is an element of `B` that plays the role of a primitive `n`-th root of unity.
+* `is_primitive_root.power_basis`: if `K` and `L` are fields such that
+  `is_cyclotomic_extension {n} K L` and `ne_zero (↑n : K)`, then `is_primitive_root.power_basis`
+  gives a K-power basis for L given a primitive root `ζ`.
+* `is_primitive_root.embeddings_equiv_primitive_roots`: the equivalence between `L →ₐ[K] A`
+  and `primitive_roots n A` given by the choice of `ζ`.
+
+## Main results
+* `is_cyclotomic_extension.zeta_primitive_root`: if `is_domain B` and `ne_zero (↑n : B)`, then
+  `zeta n A B` is a primitive `n`-th root of unity.
+* `is_cyclotomic_extension.finrank`: if `irreducible (cyclotomic n K)` (in particular for
+  `K = ℚ`), then the `finrank` of a cyclotomic extension is `n.totient`.
+* `is_primitive_root.norm_eq_one`: If `K` is linearly ordered (in particular for `K = ℚ`), the norm
+  of a primitive root is `1` if `n` is odd.
+* `is_primitive_root.sub_one_norm_eq_eval_cyclotomic`: If `irreducible (cyclotomic n K)`, then
+  the norm of `ζ - 1` is `eval 1 (cyclotomic n ℤ)`.
+
+## Implementation details
+`zeta n A B` is defined as any root of `cyclotomic n A` in `B`, that exists because of
+`is_cyclotomic_extension {n} A B`. It is not true in general that it is a primitive `n`-th root of
+unity, but this holds if `is_domain B` and `ne_zero (↑n : B)`.
+
+`zeta n A B` is defined using `Exists.some`, which means we have zero control about what it is.
+For example, in normal mathematics we can demand that `(zeta p ℤ ℤ[ζₚ] : ℚ(ζₚ))` is equal to
+`zeta p ℚ ℚ(ζₚ)`, as we are just choosing "an arbitrary" one and we can internally specify that they
+should agree. This is NOT the case for Lean (without much difficulty), and indeed it is independent
+of Lean whether these two are equal or not. Therefore, whenever possible, we specialize results to
+primitive roots, and only at the "final step", when we need to provide a primitive root as evidence
+that stuff works, should we then use zeta.
```

Comment:
```suggestion
`zeta n A B` is defined using `exists.some`, which means we cannot control it.
For example, in normal mathematics, we can demand that `(zeta p ℤ ℤ[ζₚ] : ℚ(ζₚ))` is equal to
`zeta p ℚ ℚ(ζₚ)`, as we are just choosing "an arbitrary primitive root" and we can internally specify that
our choices agree. This is not the case here, and it is indeed impossible to prove that these two are equal.
Therefore, whenever possible, we prove our results for any primitive root, and only at the "final step", when
we need to provide an "explicit" primitive root, we then use zeta.
```
Let's try to look a little more professional :)

Reply:
Many thanks!